students/BrianB/lesson07/html_render.py

```diff
@@ -0,0 +1,225 @@
+
+    def render(self, out_file, indent="", ind_count=0):
+        # loop through the list of contents
+        out_file.write("<{}>\n".format(self.tag))
+        for content in self.contents:
+            try:
+                content.render(out_file)
+            except AttributeError:
+                out_file.write(content)
+            out_file.write("\n")
+        out_file.write("</{}>\n".format(self.tag))
+
+
+class Html(Element):
+    """
+    Framework for 'html' tag, inherits from Element.
+    Overrides render method and tag class attribute.
+    """
+    tag = 'html'
+
+    def render(self, out_file, indent="", ind_count=0):
+        """
+        I was unable to follow the instructions beyond
+        mid-Step 4, this was borrowed from an alt source
+        """
+        self.indent = indent
+        out_file.write(f'<!DOCTYPE html>\n')
+        super().render(out_file, indent)
+
+
+class Body(Element):
+    tag = "body"
+
+
+class P(Element):
+    tag = "p"
+
+
+class Head(Element):
+    tag = "head"
+
+
+class Li(Element):
+    """
+    I was unable to follow the instructions beyond
+    mid-Step 4, this was borrowed from an alt source
+    """
+    tag = 'li'
+
+
+class Ul(Element):
+    """
+    I was unable to follow the instructions beyond
+    mid-Step 4, this was borrowed from an alt source
+    """
+    tag = 'ul'
+
+    def __init__(self, content=None, **kwargs):
+        """
+        I was unable to follow the instructions beyond
+        mid-Step 4, this was borrowed from an alt source
+        """
+        if content: raise TypeError
+
+        self.contents = []
+        self.attributes = kwargs
+
+
+class OneLineTag(Element):
+
+    def render(self, out_file):
+        # loop through the list of contents
+        out_file.write("<{}>".format(self.tag))
+        out_file.write(self.contents[0])
+        out_file.write("</{}>\n".format(self.tag))
+
+    def append(self, content):
+        raise NotImplementedError
+
+
+class H(OneLineTag):
+    """
+    I was unable to follow the instructions beyond
+    mid-Step 4, this was borrowed from an alt source
+    """
+    tag = 'h'
+
+    def __init__(self, level, content=None, **kwargs):
+        """
+        I was unable to follow the instructions beyond
+        mid-Step 4, this was borrowed from an alt source
+        """
+        try:
+            int(level)
+        except ValueError:
+            raise ValueError
+
+        self.tag = f'h{level}'
+        super().__init__(content, **kwargs)
+
+
+class Title(OneLineTag):
+    tag = "title"
+
+
+class A(OneLineTag):
+    """
+    I was unable to follow the instructions beyond
+    mid-Step 4, this was borrowed from an alt source
+    """
+    tag = 'a'
+
+    def __init__(self, link, content=None, **kwargs):
+        """
+        I was unable to follow the instructions beyond
+        mid-Step 4, this was borrowed from an alt source
+        """
+        if not (content and link): raise TypeError
```

Comment:
again, always add a message when you raise an Exception.

and do you need to have some content in an anchor tag? I'm not sure it it's required.